### ISSUE TITLE
question: how the predict method works when the activation function for the output layer is Softmax?

### ISSUE THREAD

**Issue by ivan-marroquin** [itdxer/neupy#191]
Hi,

I implemented a MLP network with Softmax as activation function in the output layer. The network predicts three classes.

Tracking the code, in the attached script, I noticed the following:
- Matrix multiplication: Input data (801,4) x Weights hidden layer (4,8) + bias => gives Output_1 (801,8)
- Matrix multiplication: Output_1 (801,4) x Weights output layer (8,3) => gives Output_2 (801,3)

For the final output (Output_2), I also noticed that the addition of each tuple gives 1. This is the expected behavior since Softmax is a normalized exponential function. 

So, there is something done in between "Output_1 x Weights output layer"  and generate "Output_2". Could someone explain to me how this is done? 

Many thanks,

Ivan
[Test_script_data.zip](https://github.com/itdxer/neupy/files/1243592/Test_script_data.zip)

**Comment by itdxer** [comment]
Hi,
The softmax layer after matrix multiplication applies softmax function to the output: https://en.wikipedia.org/wiki/Softmax_function
Softmax function converts input vector to the discrete probability distribution, meaning that all sum of probabilities will be equal to 1 (as you already know).

**Comment by ivan-marroquin** [comment]
Hi there,

Thanks for the information. I would like to ask you a follow up question.

In the first script that I sent, if you replace the command in line 405 by these lines of code:
facies_prob_density= best_model.predict(xtest_data[773,:])
print(facies_prob_density)
sys.exit()

The output is [ 0.55090469  0.19741563  0.25167969]. They represent the probability of three lithofacies

I am working on a script that will help me to do a reverse mapping (from neural network output back to input values). To that end, I need to understand how the predict method is implemented in Neupy. Thanks to your advice, I come up with a new script (see attachment, also included all necessary input files) that makes use of :
- The weights from the hidden and output layers
- The bias from the hidden and output layers
- The activation function for the hidden and output layers

Then, I used the same inputs (as shown above: xtest_data[773,:]) in the new script. The output is: 
Best probability lithofacies 1 0.5399078726768494
Best probability lithofacies 2 0.20537355542182922
Best probability lithofacies 3 0.25471851229667664

I checked the source code to understand the way how the activation functions are defined, and I can't explain the numerical difference.

Any suggestions?

Many thanks,

Ivan


[Test_Optimization.zip](https://github.com/itdxer/neupy/files/1247109/Test_Optimization.zip)

**Comment by itdxer** [comment]
I'm sorry. I'm not sure that I understood your question. Can you clarify it?

If you are looking for functions that apply matrix multiplications and activation function - then you have to check functionality of the specific layer. Layers that you use are located here: https://github.com/itdxer/neupy/blob/master/neupy/layers/activations.py

**Comment by ivan-marroquin** [comment]
Hi,

I went to check the code in Neupy and Theano, and I believe that I found where I was doing a  mistake. 

Thanks for all,

Ivan
